File: vivo-c2rust/01-Primary/test-without-alloc/test-sortedarray.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <time.h>
#include "compare-int.h"
#include "sortedarray.h"
#define TEST_SIZE 20

#define TEST_ARRAY {10, 12, 12, 1, 2, 3, 6, 7, 2, 23, 13, 23, 23, 34, 31, 9,\
   	21, -2, -12, -4}

#define TEST_REMOVE_EL 15

#define TEST_REMOVE_RANGE 7

#define TEST_REMOVE_RANGE_LENGTH 4

void check_sorted_prop(SortedArray *sortedarray)
{
	unsigned int i;
	for (i = 1; i < sortedarray_length(sortedarray); i++) {
		assert(int_compare(
		                   sortedarray_get(sortedarray, i-1),
						   sortedarray_get(sortedarray, i)) <= 0);
	}
}
void free_sorted_ints(SortedArray *sortedarray)
{
	unsigned int i;
	for (i = 0; i < sortedarray_length(sortedarray); i++) {
		int *pi = (int*) sortedarray_get(sortedarray, i);
		free(pi);
	}

	sortedarray_free(sortedarray);
}
SortedArray *generate_sortedarray_equ(SortedArrayEqualFunc equ_func)
{
	/* generate a sorted array of length TEST_SIZE, filled with random 
	   numbers. */
	SortedArray *sortedarray;
	unsigned int i;

	int array[TEST_SIZE] = TEST_ARRAY;

	sortedarray = sortedarray_new(0, equ_func, int_compare);

	for (i = 0; i < TEST_SIZE; ++i) {
		int *pi = malloc(sizeof(int));
		*pi = array[i];
		sortedarray_insert(sortedarray, pi);
	}

	return sortedarray;
}
SortedArray *generate_sortedarray(void)
{
	return generate_sortedarray_equ(int_equal);
}
void test_sortedarray_new_free(void)
{
	SortedArray *sortedarray;

	/* test normal */
	sortedarray = sortedarray_new(0, int_equal, int_compare);
	assert(sortedarray != NULL);
	sortedarray_free(sortedarray);

	/* freeing null */
	sortedarray_free(NULL);

	/* low memory */
	}
void test_sortedarray_insert(void)
{
	SortedArray *sortedarray = generate_sortedarray();
	unsigned int i;

	/* insert a few random numbers, then check if everything is sorted */
	for (i = 0; i < 20; i++) {
		int i = (int) (((float) rand())/((float) RAND_MAX) * 100);
		int *pi = malloc(sizeof(int));
		*pi = i;
		sortedarray_insert(sortedarray, pi);
	}

	check_sorted_prop(sortedarray);
	free_sorted_ints(sortedarray);
}
void test_sortedarray_remove(void)
{
	SortedArray *sortedarray = generate_sortedarray();

	/* remove index 24 */
	int *ip = (int*) sortedarray_get(sortedarray, TEST_REMOVE_EL + 1);
	int i = *ip;
	free((int*) sortedarray_get(sortedarray, TEST_REMOVE_EL));
	sortedarray_remove(sortedarray, TEST_REMOVE_EL);
	assert(*((int*) sortedarray_get(sortedarray, TEST_REMOVE_EL)) == i);

	check_sorted_prop(sortedarray);
	free_sorted_ints(sortedarray);
}
void test_sortedarray_remove_range(void)
{
	SortedArray *sortedarray = generate_sortedarray();

	/* get values in test range */
	int new[TEST_REMOVE_RANGE_LENGTH];
	unsigned int i;
	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
		new[i] = *((int*) sortedarray_get(sortedarray, TEST_REMOVE_RANGE + 
		                                    TEST_REMOVE_RANGE_LENGTH + i));
	}
	
	/* free removed elements */
	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
		free((int*) sortedarray_get(sortedarray, TEST_REMOVE_RANGE + i));
	}

	/* remove */
	sortedarray_remove_range(sortedarray, TEST_REMOVE_RANGE, 
			TEST_REMOVE_RANGE_LENGTH);
	
	/* assert */
	for (i = 0; i < TEST_REMOVE_RANGE_LENGTH; i++) {
		assert(*((int*) sortedarray_get(sortedarray, TEST_REMOVE_RANGE + i)) == 
		                                                                new[i]);
	}

	check_sorted_prop(sortedarray);
	free_sorted_ints(sortedarray);
}
void test_sortedarray_index_of(void) {
	SortedArray *sortedarray = generate_sortedarray();

	unsigned int i;
	for (i = 0; i < TEST_SIZE; i++) {
		int r = sortedarray_index_of(sortedarray, 
		                sortedarray_get(sortedarray, i));
		assert(r >= 0);
		assert(*((int*) sortedarray_get(sortedarray,(unsigned int) r)) == 
		        *((int*) sortedarray_get(sortedarray, i)));
	}
	
	free_sorted_ints(sortedarray);
}
static int ptr_equal(SortedArrayValue v1, SortedArrayValue v2) {
	return v1 == v2;
}
void test_sortedarray_index_of_equ_key(void)
{
	/* replace equal function by function which checks pointers */
	SortedArray *sortedarray = generate_sortedarray_equ(ptr_equal);
	unsigned int i;

	/* check if all search value return the same index */
	for (i = 0; i < TEST_SIZE; i++) {
		int r = sortedarray_index_of(sortedarray, 
		                             sortedarray_get(sortedarray, i));
		assert(r >= 0);
		assert(i == (unsigned int) r);
	}

	free_sorted_ints(sortedarray);
}
void test_sortedarray_get(void) {
	unsigned int i;

	SortedArray *arr = generate_sortedarray();

	for (i = 0; i < sortedarray_length(arr); i++) {
		assert(sortedarray_get(arr, i) == sortedarray_get(arr, i));
		assert(*((int*) sortedarray_get(arr, i)) == 
		       *((int*) sortedarray_get(arr, i)));
	}

	free_sorted_ints(arr);
}

int main(int argc, char *argv[])
{
	return 0;
}

File: Output/primary/src/test_sortedarray.rs
use crate::compare_int::*;
use crate::sortedarray::*;
use crate::translation_utils::*;macro_rules! test_size { () => { 20 }; }
macro_rules! test_array { () => { arr![10, 12, 12, 1, 2, 3, 6, 7, 2, 23, 13, 23, 23, 34, 31, 9, 21, -2, -12, -4] }; }
macro_rules! test_remove_el { () => { 15 }; }
macro_rules! test_remove_range { () => { 7 }; }
macro_rules! test_remove_range_length { () => { 4 }; }
fn check_sorted_prop(mut sortedarray: Unowned<SortedArray<Ptr<i32>>>) {
    let mut i: u32;
    c_for!(i = 1; i < sortedarray_length(sortedarray); i += 1; {
        assert!(int_compare(sortedarray_get(sortedarray, i - 1), sortedarray_get(sortedarray, i)) <= 0);
    });
}
fn free_sorted_ints(mut sortedarray: Owned<SortedArray<Ptr<i32>>>) {
    let mut i: u32;
    c_for!(i = 0; i < sortedarray_length(sortedarray.unowned()); i += 1; {
        let mut pi: Ptr<i32> = sortedarray_get(sortedarray.unowned(), i);
        c_free!(pi);
    });
    sortedarray_free(sortedarray.unowned());
}
fn generate_sortedarray_equ(mut equ_func: SortedArrayEqualFunc<Ptr<i32>>) -> Owned<SortedArray<Ptr<i32>>> {
    let mut sortedarray: Owned<SortedArray<Ptr<i32>>>;
    let mut i: u32;
    let array: Array<i32, {test_size!()}> = test_array!();
    sortedarray = sortedarray_new(0, equ_func, func!(int_compare));
    c_for!(i = 0; i < test_size!(); i += 1; {
        let mut pi: Ptr<i32> = c_malloc!(c_sizeof!(i32));
        *pi = array[i];
        sortedarray_insert(sortedarray.unowned(), pi);
    });
    sortedarray
}
fn generate_sortedarray() -> Owned<SortedArray<Ptr<i32>>> {
    generate_sortedarray_equ(func!(int_equal))
}
#[test]
fn test_sortedarray_new_free() {
    let mut sortedarray: Owned<SortedArray<Ptr<i32>>>;
    sortedarray = sortedarray_new(0, func!(int_equal), func!(int_compare));
    assert!(sortedarray != null!());
    sortedarray_free(sortedarray.unowned());
    sortedarray_free::<Ptr<i32>>(null!());
}
#[test]
fn test_sortedarray_insert() {
    let mut sortedarray: Owned<SortedArray<Ptr<i32>>> = generate_sortedarray();
    let mut i: u32;
    c_for!(i = 0; i < 20; i += 1; {
        let i: i32 = unsafe { (rand() as f32 / RAND_MAX as f32 * 100.0) as i32 };
        let mut pi: Ptr<i32> = c_malloc!(c_sizeof!(i32));
        *pi = i;
        sortedarray_insert(sortedarray.unowned(), pi);
    });
    check_sorted_prop(sortedarray.unowned());
    free_sorted_ints(sortedarray);
}
pub fn rb_tree_remove<T: GenericValue>(mut tree: Unowned<RBTree<T>>, key: RBTreeKey<T>) -> i32 {
    let mut node = rb_tree_lookup_node(tree, key);
    if node == null!() {
        return 0;
    }
    rb_tree_remove_node(tree, node);
    return 1;
}
pub fn rb_tree_remove<T: GenericValue>(mut tree: Unowned<RBTree<T>>, key: RBTreeKey<T>) -> i32 {
    let mut node = rb_tree_lookup_node(tree, key);
    if node == null!() {
        return 0;
    }
    rb_tree_remove_node(tree, node);
    return 1;
}
#[test]
fn test_sortedarray_index_of() {
    let mut sortedarray: Owned<SortedArray<Ptr<i32>>> = generate_sortedarray();
    let mut i: u32;
    c_for!(i = 0; i < test_size!(); i += 1; {
        let mut r: i32 = sortedarray_index_of(sortedarray.unowned(), sortedarray_get(sortedarray.unowned(), i));
        assert!(r >= 0);
        assert!(*sortedarray_get(sortedarray.unowned(), r as u32) == *sortedarray_get(sortedarray.unowned(), i));
    });
    free_sorted_ints(sortedarray);
}
fn ptr_equal<T: GenericValue>(mut v1: SortedArrayValue<T>, mut v2: SortedArrayValue<T>) -> i32 {
    (v1 == v2) as i32
}
#[test]
fn test_sortedarray_index_of_equ_key() {
    let mut sortedarray: Owned<SortedArray<Ptr<i32>>> = generate_sortedarray_equ(func!(ptr_equal));
    let mut i: u32;
    c_for!(i = 0; i < test_size!(); i += 1; {
        let mut r: i32 = sortedarray_index_of(sortedarray.unowned(), sortedarray_get(sortedarray.unowned(), i));
        assert!(r >= 0);
        assert!(i == r as u32);
    });
    free_sorted_ints(sortedarray);
}
#[test]
fn test_sortedarray_get() {
    let mut i: u32;
    let mut arr: Owned<SortedArray<Ptr<i32>>> = generate_sortedarray();
    c_for!(i = 0; i < sortedarray_length(arr.unowned()); i += 1; {
        assert!(sortedarray_get(arr.unowned(), i) == sortedarray_get(arr.unowned(), i));
        assert!(*sortedarray_get(arr.unowned(), i) == *sortedarray_get(arr.unowned(), i));
    });
    free_sorted_ints(arr);
}
